code/modules/power/tesla/energy_ball.dm

```diff
@@ -0,0 +1,203 @@
+	var/closest_atom
+	for(var/atom/A in oview(source, zap_range))
+		if(istype(A, /obj/machinery/power/tesla_coil))
+			var/obj/machinery/power/tesla_coil/C = A
+			if(C.being_shocked)
+				continue
+			tesla_coils.Add(C)
+			continue
+		if(istype(A, /obj/machinery/power/grounding_rod))
+			var/obj/machinery/power/grounding_rod/R = A
+			grounding_rods.Add(R)
+			continue
+		if(istype(A, /obj/machinery))
+			var/obj/machinery/M = A
+			if(is_type_in_list(M, blacklisted_tesla_types))
+				continue
+			if(M.being_shocked)
+				continue
+			potential_machine_zaps.Add(M)
+			continue
+		if(istype(A, /obj/structure))
+			var/obj/structure/M = A
+			if(is_type_in_list(M, blacklisted_tesla_types))
+				continue
+			if(M.being_shocked)
+				continue
+			potential_structure_zaps.Add(M)
+			continue
+		if(istype(A, /mob/living))
+			var/mob/living/L = A
+			if(L.stat == DEAD)
+				continue
+			if(is_type_in_list(L, blacklisted_tesla_types))
+				continue
+			potential_mob_zaps.Add(L)
+			continue
+	closest_atom = get_closest_atom(/obj/machinery/power/tesla_coil, tesla_coils, source)
+	if(closest_atom && istype(closest_atom, /obj/machinery/power/tesla_coil))
+		var/obj/machinery/power/tesla_coil/C = closest_atom
+		source.Beam(C,icon_state="lightning[rand(1,12)]",icon='icons/effects/effects.dmi',time=5)
+		C.tesla_act(power)
+		return
+	if(!closest_atom)
+		closest_atom = get_closest_atom(/obj/machinery/power/grounding_rod, grounding_rods, source)
+		if(closest_atom && istype(closest_atom, /obj/machinery/power/grounding_rod))
+			var/obj/machinery/power/grounding_rod/R = closest_atom
+			source.Beam(R,icon_state="lightning[rand(1,12)]",icon='icons/effects/effects.dmi',time=5)
+			R.tesla_act(power)
+			return
+	if(!closest_atom)
+		closest_atom = get_closest_atom(/mob/living, potential_mob_zaps, source)
+		if(closest_atom && istype(closest_atom, /mob/living))
+			var/mob/living/L = closest_atom
+			var/shock_damage = Clamp(round(power/400), 10, 90) + rand(-5,5)
+			source.Beam(L,icon_state="lightning[rand(1,12)]",icon='icons/effects/effects.dmi',time=5)
+			L.electrocute_act(shock_damage, source, 1, tesla_shock = 1)
+			if(istype(L, /mob/living/silicon))
+				var/mob/living/silicon/S = L
+				S.emp_act(2)
+			return
+	if(!closest_atom)
+		closest_atom = get_closest_atom(/obj/machinery, potential_machine_zaps, source)
+		if(closest_atom)
+			var/obj/machinery/M = closest_atom
+			source.Beam(M,icon_state="lightning[rand(1,12)]",icon='icons/effects/effects.dmi',time=5)
+			M.tesla_act(power)
+			if(prob(85))
+				M.emp_act(2)
+			else
+				if(prob(50))
+					M.ex_act(3)
+				else
+					if(prob(90))
+						M.ex_act(2)
+					else
+						M.ex_act(1)
+			return
+	if(!closest_atom)
+		closest_atom = get_closest_atom(/obj/structure, potential_structure_zaps, source)
+		if(closest_atom)
+			var/obj/structure/S = closest_atom
+			source.Beam(S,icon_state="lightning[rand(1,12)]",icon='icons/effects/effects.dmi',time=5)
+			S.tesla_act(power)
+			return
```

Comment:
This proc is awful, but I can't think of any way to improve it without compromising on performance (which matters more in this case), and it's pretty well contained and modular, so I think it can stay.